lib/private/Preview/Imaginary.php

```diff
@@ -0,0 +1,115 @@
+<?php
+/**
+ * @copyright Copyright (c) 2020, Nextcloud, GmbH.
+ *
+ * @author Vincent Petry <vincent@nextcloud.com>
+ * @author Carl Schwan <carl@carlschwan.eu>
+ *
+ * @license AGPL-3.0-or-later
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program. If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OC\Preview;
+
+use OCP\Files\File;
+use OCP\Http\Client\IClientService;
+use OCP\IConfig;
+use OCP\IImage;
+
+use OC\StreamImage;
+use Psr\Log\LoggerInterface;
+
+class Imaginary extends ProviderV2 {
+	/** @var IConfig */
+	private $config;
+
+	/** @var IClientService */
+	private $service;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	public function __construct(array $config) {
+		parent::__construct($config);
+		$this->config = \OC::$server->get(IConfig::class);
+		$this->service = \OC::$server->get(IClientService::class);
+		$this->logger = \OC::$server->get(LoggerInterface::class);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function getMimeType(): string {
+		return '/image\/.*/';
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
+		$maxSizeForImages = $this->config->getSystemValue('preview_max_filesize_image', 50);
+		$size = $file->getSize();
+
+		if ($maxSizeForImages !== -1 && $size > ($maxSizeForImages * 1024 * 1024)) {
+			return null;
+		}
+
+		$imaginaryUrl = $this->config->getSystemValueString('preview_imaginary_url', 'invalid');
+		if ($imaginaryUrl === 'invalid') {
+			$this->logger->error('Imaginary preview provider is enabled, but no url is configured. Please provide the url of your imaginary server to the \'preview_imaginary_url\' config variable.');
+			return null;
+		}
+		$imaginaryUrl = rtrim($imaginaryUrl, '/');
+
+		// Object store
+		$stream = $file->fopen('r');
+
+		$httpClient = $this->service->newClient();
+		$parameters = http_build_query([
+			'width' => $maxX,
+			'height' => $maxY,
+			'stripmeta' => 'true',
+			'type' => 'jpeg',
+		]);
+
+		try {
+			$response = $httpClient->post(
+				$imaginaryUrl . '/fit?' . $parameters, [
+					'stream' => true,
+					'content-type' => $file->getMimeType(),
+					'body' => $stream,
+					'nextcloud' => ['allow_local_address' => true],
+				]);
```

Comment:
```suggestion
		$parameters = [
			'width' => $maxX,
			'height' => $maxY,
			'stripmeta' => 'true',
			'type' => 'jpeg',
		];

		try {
			$response = $httpClient->post(
				$imaginaryUrl . '/fit', [
					'query' => $parameters,
					'stream' => true,
					'content-type' => $file->getMimeType(),
					'body' => $stream,
					'nextcloud' => ['allow_local_address' => true],
				]);
```

I’m pretty sure you can do it like this, query entry is not documented for post in our http client class but I think guzzle will accept it on any request, the opposite would be weird.

Reply:
it works